### ISSUE TITLE
Failing on Erista

### ISSUE THREAD

**Issue by classic-gentleman** [KazushiMe/Switch-OC-Suite#20]
Relevant log:

[2022-05-06 02:57:37.604] [!] ASSERT RESULT_OK: clkrstSetClockRate
RC: [0x8c5] 0197-0004
in /workspace/switch/sys-clk/sysmodule/src/clocks.cpp:227
[2022-05-06 02:57:37.609] Exit
[2022-05-06 03:03:09.005] === sys-clk1.0.1-5-gdcd0d5d ===

Latest everything, injecting fusee.bin. When I don't load the kip the same sys-clk works just fine. I'm mainly interested in memory OC.

**Comment by classic-gentleman** [comment]
It is the latest `development` commit. The only difference to `master` is disabling of logs.

https://github.com/retronx-team/sys-clk/commit/b38bac52bcfbf22ee7ec0a76d3b91f3b03b0d741

I'll fetch a full log later today but this truly is the only relevant part... Per the source code file and line the assertion refers to, it's failing to set the frequency somehow.

**Comment by KazushiMe** [comment]
Added `ASSERT_RESULT_OK(rc, "clkrstSetClockRate, module: %s, hz: %lu", sysclkFormatModule(module, false), hz);` to Line 227: [exefs.nsp.zip](https://github.com/KazushiMe/Switch-OC-Suite/files/8652371/exefs.nsp.zip). See what module and clock rate could throw this exception.

**Comment by classic-gentleman** [comment]
```
[2022-05-10 00:43:44.005] === sys-clk dcd0d5d-dirty ===
[2022-05-10 00:43:56.775] Ready
[2022-05-10 00:43:56.794] [mgr] sys-clk status: enabled
[2022-05-10 00:43:56.972] [mgr] TitleID change: 0100000000001000
[2022-05-10 00:43:57.172] [!] ASSERT_RESULT_OK: clkrstSetClockRate, module: mem, hz: 1331200000
RC: [0x8c5] 0197-0004
in/Volumes/RAM/sys-clk/sysmodule/src/clocks.cpp:227
[2022-05-10 00:43:57.178] Exit
```

Interesting...

**Comment by KazushiMe** [comment]
Not sure why this happens as loader.kip only replaces mem frequency 1600000 khz with 1862400 khz for erista by default. 

Some have reported that fusee.bin is not compatible custom kips, only hekate fss0 works. 

[loader_1600_1.3.2.kip.zip](https://github.com/KazushiMe/Switch-OC-Suite/files/8656241/loader_1600_1.3.2.kip.zip) disables mem clock patching for erista.

**Comment by KazushiMe** [comment]
[exefs.nsp.zip](https://github.com/KazushiMe/Switch-OC-Suite/files/8666871/exefs.nsp.zip) would ignore exception from `clkrstSetClockRate()`.

I don't have erista consoles to test. Will keep this issue opened and wait for reports from other users.

**Comment by classic-gentleman** [comment]
> [exefs.nsp.zip](https://github.com/KazushiMe/Switch-OC-Suite/files/8666871/exefs.nsp.zip) would ignore exception from `clkrstSetClockRate()`.
> 
> I don't have erista consoles to test. Will keep this issue opened and wait for reports from other users.

Well... This works! No assertions and things are seemingly working as they would on stock sys-clk. But I'm not seeing a bump to 1862 on the memory clock when setting to 1600. As a matter of fact this is setting 1600 on boot (I've cleared my config.ini) instead of 1331. Did you mix this with the previous change and the end result is avoiding overclocking memory above 1600 on Erista?

**Comment by llyama** [comment]
I was having this exact same problem. I fixed it by removing `bootloader/sys/libsys_minerva.bso` as per the readme.

Booting with fusee or with hekate with the minerva file causes this issue. Hekate without minerva gives warnings on boot about the missing file but the overlock works and upclocks the ram to 1862.4MHz once 1600 is requested.

**Comment by KazushiMe** [comment]
Noted. Patching minerva module is possible if you don't feel like rebuilding hekate.
`cd` to sdcard and run in `python3`:
```python3
rate_khz_patch = 1862_400 #Erista only
loc = "./bootloader/sys/libsys_minerva.bso" # path to libsys_minerva.bso

rate_khz_1600 = 1600_000
min_volt = 887
magic = [rate_khz_1600, min_volt, rate_khz_patch]
magic_bytes = [i.to_bytes(4, 'little') for i in magic]
offsets = []
with open(loc, "rb") as file:
    offset = -1
    while True:
        file.seek(offset + 1)
        buf = file.read()
        pos = buf.find(magic_bytes[0])
        if pos == -1:
            break
        offset += 1 + pos
        file.seek(offset + 4)
        if file.read()[0:4] == magic_bytes[1]:
            offsets.append(offset)
print(f"Found {len(offsets)} occurrence(s):", [hex(offset) for offset in offsets])
with open(loc, "rb+") as file:
    for offset in offsets:
        file.seek(offset)
        file.write(magic_bytes[2])
print("Done!")
```

**Comment by Pinguin762** [comment]
> Booting with fusee or with hekate with the minerva file causes this issue. Hekate without minerva gives warnings on boot about the missing file but the overlock works and upclocks the ram to 1862.4MHz once 1600 is requested.

Same for me. Works whithout minerva, not with fusee or with minerva

> Noted. Patching minerva module is possible if you don't feel like rebuilding hekate.
> `cd` to sdcard and run in `python3`:

Maybe I'm doing something wrong. When patching minerva Hekate stops giving warning but won't boot in Atmosphère just a blank screen after the atmos logo, fusee boots but won't work sys-clk either. Using official sys-clk

![image](https://user-images.githubusercontent.com/40664560/183518502-c11043c8-7e84-43ef-9491-c71e52ee4894.png)

Maybe I'm justing missing something patching idk

**Comment by llyama** [comment]
I had the same thing happen. Patching libsys_minerva.bso isn't enough on it's own. You also need to patch the main Hekate bin file.

The main bin has some minerva code that references 1600mhz. By patching the enum defined in [minerva.h](https://github.com/CTCaer/hekate/blob/master/bdk/mem/minerva.h#L58) to make the highest value 1862400 instead of 1600000 makes it all work.

Ideally Hekate needs recompiling with these changes properly thought out. Instead... I just hex edited the binary to change the value.

You need to change `00 6A 18 00` to `00 6B 1C 00` at offset `0x7A45`.

Obviously this only works with the release build of Hekate 5.8.0.

**Comment by Pinguin762** [comment]
> You need to change `00 6A 18 00` to `00 6B 1C 00` at offset `0x7A45`.

That worked, thanks for the help

**Comment by Nazosan** [comment]
Seems the issue is still present in the latest versions of everything too.  (I too am on Erista.)  I tried using the Python patch script above and manually changing the first occurrence of 00 6A 18 00 (at 0x7AF1) to 00 6B 1C 00 in the update.bin but the sys-clk module is still failing to load.  I also tried the second (only other) occurrence at 0xDD23 just to be sure, but it also didn't work.  I even tried changing both.  (I just load the fusee.bin payload for simplicity of making updates much easier if that makes any difference btw.)   I'm not quite sure what Minerva is or what the consequences of removing that module may be.

Clean log (not much difference from the above):
```
[2022-10-22 18:36:27.006] === sys-clk 1.0.2 ===
[2022-10-22 18:36:35.244] Ready
[2022-10-22 18:36:35.269] [mgr] sys-clk status: enabled
[2022-10-22 18:36:35.340] [mgr] TitleID change: 0100000000001000
[2022-10-22 18:36:35.387] [!] ASSERT_RESULT_OK: clkrstSetClockRate
  RC: [0x8c5] 0197-0004
  in /workspace/switch/sys-clk/sysmodule/src/clocks.cpp:227
[2022-10-22 18:36:35.422] Exit
```

Perhaps the simplest long term solution might be just to have a custom sys-clk for Minerva?

**Comment by KazushiMe** [comment]
Fusee and Hekate both train mem @ 1600 MHz. Pcv might not work if 1600 MHz table is absent.

Lowest frequency 40.8 MHz is discarded to make room for 1600 MHz, see if [this](https://github.com/KazushiMe/Switch-OC-Suite/files/9845798/1.4.0_loader.kip.zip) works.

**Comment by Nazosan** [comment]
It sounds like you may be on to the right sort of idea there.  I don't think 40.8 is going to be seriously used for anything.  Anyway, sys-clk loads when I use that kip, but there is no 1862.4 MHz option for the memory in the overlay or the manager (well, I guess that's third party.)  Actually, the options are extremely limited with only "do not override," 665, 800, 1065, 1331, and 1600 in the list for either.  I tried putting it in the INI file directly but it didn't work.  I actually had it on the home screen title just to test quickly and easily and it got stuck in bootup on a blank black screen that way.  I rebooted it in handheld (I only changed the docked memory speed, not handheld) and then docked it and it set 1600MHz despite actually saying 1862 in the profile on the display.

**Comment by KazushiMe** [comment]
Fixed: https://github.com/KazushiMe/Switch-OC-Suite/releases/tag/1.4.0.
Loader.kip uploaded along with modified sys-clk for Erista. No manual patching to fusee or hekate required.

Also @Nazosan your config.ini is broken, delete it and use sys-clk overlay / manager to configure instead.

**Comment by Nazosan** [comment]
The INI file just had a couple of minor errors.  I corrected them after I saw the log already.  Most of the game profiles were added via copy/paste from a large public list and the thing that came up over and over was a minor typo someone had in there, which I fixed.  The first error was just from testing something a while back and has been removed.

Anyway, this seems to work.  The 1862.4 option at least declares to correctly set that rate.  (I guess I'd have to benchmark to completely confirm.)  I just thought I'd mention in case it matters, but the Erista-only sys-clk is also showing 1996 and 2131 speeds as options.  I'm not sure if it's intended those be available when not using raised voltages and etc to do true overclocking.  I didn't test them though.
